### PR TITLE
ENH: Implement EA unary ops

### LINKED ISSUE #23087: Implement/dispatch unary operations for ExtensionArrays
xref https://github.com/pandas-dev/pandas/pull/22325/files#r220580694

unary ops are not implemented by EAs and also not dispatched to EAs:

```
In [161]: s = pd.Series([1, 2, 3], dtype='Int64')

In [162]: -s
Out[162]: 
0    -1
1    -2
2    -3
dtype: object
```

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] closes #23087
- [x] tests added / passed
    - [x] Array
    - [x] Series
    - [x] DataFrame
    - [x] Panel
    - [x] Sparse
    - [x] Extension (Decimal)
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Changed to apply unary funcs per blocks. As a result, unexpected dtype changes are also fixed (as below).
```
# Current master
(-pd.DataFrame({'a': [1], 'b': [1.1]})).dtypes
# a    float64
# b    float64
# dtype: object

# After the PR
(-pd.DataFrame({'a': [1], 'b': [1.1]})).dtypes
# a      int64
# b    float64
# dtype: object
```

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -1250,33 +1250,17 @@ def _indexed_same(self, other):
...
-        if is_bool_dtype(values):
-            arr = operator.inv(values)
-        elif (is_numeric_dtype(values) or is_timedelta64_dtype(values)
-                or is_object_dtype(values)):
-            arr = operator.neg(values)
-        else:
-            raise TypeError("Unary negative expects numeric dtype, not {}"
-                            .format(values.dtype))
-        return self.__array_wrap__(arr)
+        new_data = self._data.apply(operator.neg)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/generic.py]: We've gotten arithmetic/comparison ops out of `internals`.  Haven't done anything to unary ops.

[the remark this answers is not in the document]
Comment (sinhrks) [reply pandas/core/generic.py]: Thx, I'm going to move the logic to `ops.py` immitating binary ops manner.

File: pandas/tests/series/test_arithmetic.py [hunk 2]
```diff
@@ -82,3 +82,109 @@ def test_pow_ops_object():
     expected = Series(b.values ** a.values, dtype=object)
 
     tm.assert_series_equal(result, expected)
+
+
+# ------------------------------------------------------------------
+# Unary
+
```
Comment (jreback) [inline pandas/tests/series/test_arithmetic.py]: I think @jbrockmendel move things like this to pandas/tests/test_arithmetic.py ?

Comment (jbrockmendel) [reply pandas/tests/series/test_arithmetic.py]: unary ops we haven't been collecting in tests/arithmetic

File: pandas/tests/series/test_arithmetic.py [hunk 3]
```diff
@@ -82,3 +82,109 @@ def test_pow_ops_object():
...
+        tm.assert_series_equal(result, exp)
+
+    @pytest.mark.parametrize('typ', ['float64', 'float32', 'float16'])
+    @pytest.mark.parametrize('op', [operator.pos, operator.neg,
+                                    operator.inv, operator.abs])
+    def test_float(self, typ, op):
+        # GH#23087
+        ser = Series([1.1, 3.1, 2.1], index=range(3), dtype=typ)
+
+        if op is operator.inv:
```
Comment (jbrockmendel) [inline pandas/tests/series/test_arithmetic.py]: simpler to do operator.inv in a separate test and leave it out of the parameterize

File: pandas/core/ops.py [hunk 4]
```diff
@@ -1156,6 +1157,82 @@ def add_flex_arithmetic_methods(cls):
...
+            new_data = {i: op(self.iloc[:, i])
+                        for i in range(len(self.columns))}
+            result = self._constructor(new_data, index=self.index, copy=False)
+            # Pin columns instead of passing to constructor for compat with
+            # non-unique columns case
+            result.columns = self.columns
+            return result.__finalize__(self)
+        else:
+            with np.errstate(all='ignore'):
+                new_data = op(self.values)
```
Comment (jorisvandenbossche) [inline pandas/core/ops.py]: Should this be `self._values` to make sure it is called on the ExtensionArray ?

File: pandas/tests/arrays/test_integer.py [hunk 5]
```diff
@@ -352,6 +354,57 @@ def test_compare_array(self, data, all_compare_operators):
         self._compare_other(s, data, op_name, other)
 
 
+# ------------------------------------------------------------------
+# Unary
+
+class TestUnaryOps(object):
+
+    @pytest.mark.parametrize('typ', ['Int64', 'Int32', 'Int16', 'Int8',
+                                     'UInt64', 'UInt32', 'UInt16', 'UInt8'])
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/test_integer.py]: there is a fixture defined for this at the top of the file you can reuse

File: pandas/core/ops.py [hunk 6]
```diff
@@ -1156,6 +1157,82 @@ def add_flex_arithmetic_methods(cls):
...
+        if self.ndim > 2:
+            with np.errstate(all='ignore'):
+                # apply op at once based on current impl
+                new_data = op(self.values)
+            result = self._constructor(new_data, self.items,
+                                       self.major_axis, self.minor_axis)
+            return result.__finalize__(self)
+        elif self.ndim == 2:
+            new_data = {i: op(self.iloc[:, i])
+                        for i in range(len(self.columns))}
```
Comment (jorisvandenbossche) [inline pandas/core/ops.py]: So now this happens column by column, while before it was done on the full 2D array of the DataFrame?

File: pandas/tests/series/test_arithmetic.py (hunk 2 above)
Comment (sinhrks) [reply pandas/tests/series/test_arithmetic.py]: Moved some tests to `test_arithmetic`. Is there any guide between `test_arithmetic` and `test_operators`?

Comment (jbrockmendel) [reply pandas/tests/series/test_arithmetic.py]: It hasn't come up enough for a policy to develop.  The first thing I'd check is internal consistency: if they are in test_foo for DataFrame, they should also be in test_foo for Series and Index.

Also, if this is behavior where we expect the behaviors of `op(index)`, `op(Series(index))` and `op(DataFrame(index))` to be analogous, they could go in something like tests/arithmetic/test_unary_ops.py

Comment (sinhrks) [reply pandas/tests/series/test_arithmetic.py]: OK. Currently, similar tests are written in either `test_arithmetic.py` or `test_operators.py`. Will submit an issue to clarify this.

File: pandas/core/internals/blocks.py [hunk 7]
```diff
@@ -2216,6 +2217,12 @@ def replace(self, to_replace, value, inplace=False, filter=None,
                                               inplace=inplace, filter=filter,
                                               regex=regex, convert=convert)
 
+    def apply(self, func, **kwargs):
+        # neg should be treated as inv
```
[the remark this answers is not in the document]
Comment (sinhrks) [reply pandas/core/internals/blocks.py]: This is commonly used in internal also. Should be deprecated in 0.24?

File: pandas/core/ops.py (hunk 6 above)
Comment (sinhrks) [reply pandas/core/ops.py]: Yes. If it should consider internal structure, may better to move logics to `internals` (as in first impl)?

Comment (jorisvandenbossche) [reply pandas/core/ops.py]: I am not sure it is needed to do it block by block, but I am mainly wondering if this would have implications. 

Eg something like 

```
In [1]: df = pd.DataFrame({'a': [1, 2, 3], 'b':[False, True, False]})

In [2]: df
Out[2]: 
   a      b
0  1  False
1  2   True
2  3  False

In [3]: -df
Out[3]: 
    a   b
0  -1   0
1  -2  -1
2  -3   0

In [4]: -df['b']
Out[4]: 
0     True
1    False
2     True
Name: b, dtype: bool
```

has a different behaviour on the full dataframe vs the column (which is not necessarily bad in this case, but we should check)

File: pandas/core/ops.py [hunk 8]
```diff
@@ -1284,6 +1285,82 @@ def add_flex_arithmetic_methods(cls):
...
+    else:
+        return operator.neg(values)
+
+
+def invert(values):
+    try:
+        return operator.inv(values)
+    except Exception:
+        # inv fails with 0 len
+        if not np.prod(values.shape):
```
Comment (jbrockmendel) [inline pandas/core/ops.py]: `values.size`?  Any more specific exception?

File: pandas/core/ops.py (hunk 6 above)
Comment (sinhrks) [reply pandas/core/ops.py]: @jorisvandenbossche Yes, the PR fixes unintended dtype changes. `pandas/tests/frame/test_arithmetic.py` has some mixed dtype tests.

```
# Current master
(-pd.DataFrame({'a': [1], 'b': [1.1]})).dtypes
# a    float64
# b    float64
# dtype: object

# After the PR
(-pd.DataFrame({'a': [1], 'b': [1.1]})).dtypes
# a      int64
# b    float64
# dtype: object
```

File: pandas/core/arrays/sparse.py [hunk 9]
```diff
@@ -1645,6 +1645,7 @@ def _add_unary_ops(cls):
         cls.__pos__ = cls._create_unary_method(operator.pos)
         cls.__neg__ = cls._create_unary_method(operator.neg)
         cls.__invert__ = cls._create_unary_method(operator.invert)
+        cls.__abs__ = cls._create_unary_method(operator.abs)
```
Comment (TomAugspurger) [inline pandas/core/arrays/sparse.py]: I think SparseArray defines `__abs__`. That can be removed now, right?

**Comment by jreback** [comment]
@sinhrks can you rebase and will see where this PR is.

File: pandas/core/arrays/base.py [hunk 10]
```diff
@@ -883,3 +878,67 @@ def _create_arithmetic_method(cls, op):
     @classmethod
     def _create_comparison_method(cls, op):
         return cls._create_method(op, coerce_to_dtype=False)
+
+    @classmethod
+    def _create_unary_method(cls, op, coerce_to_dtype=True):
```
[the remark this answers is not in the document]
Comment (sinhrks) [reply pandas/core/arrays/base.py]: Created `BaseUnaryOpsTests` and added tests in  `pandas/tests/extension/decimal/test_decimal.py`.

File: pandas/core/arrays/sparse.py (hunk 9 above)
Comment (sinhrks) [reply pandas/core/arrays/sparse.py]: Correct, removed `__abs__`.

**Comment by sinhrks** [comment]
Sorry to take a long. Updated tests to meet NumPy changes. 

One remaining error in Azure is caused by network connection which looks unrelated to the change.

File: pandas/tests/frame/test_arithmetic.py [hunk 11]
```diff
@@ -634,3 +634,162 @@ def test_arith_non_pandas_object(self):
         added = pd.DataFrame(df.values + val3,
                              index=df.index, columns=df.columns)
         tm.assert_frame_equal(df.add(val3), added)
+
+
+# ------------------------------------------------------------------
+# Unary
+class TestFrameUnary(object):
+
+    @pytest.mark.parametrize('typ', ['int64', 'int32', 'int16', 'int8',
```
Comment (jreback) [inline pandas/tests/frame/test_arithmetic.py]: you can import these from pandas.conftest (we only have fixtures for the current int types, not the EA ones, though you might want to create one for combined int/EA types and use it here)

File: pandas/tests/series/test_arithmetic.py [hunk 12]
```diff
@@ -170,3 +170,150 @@ def test_ser_cmp_result_names(self, names, op):
...
+        ser.name = 'series'
+        tm.assert_series_equal(+ser, ser)
+
+    def test_invert(self):
+        ser = tm.makeStringSeries()
+        ser.name = 'series'
+        tm.assert_series_equal(-(ser < 0), ~(ser < 0))
+
+    @pytest.mark.parametrize('typ', ['int64', 'int32', 'int16', 'int8',
+                                     'uint64', 'uint32', 'uint16', 'uint8',
```
Comment (jreback) [inline pandas/tests/series/test_arithmetic.py]: same comment as above

File: pandas/tests/series/test_arithmetic.py [hunk 13]
```diff
@@ -170,3 +170,150 @@ def test_ser_cmp_result_names(self, names, op):
...
+    @pytest.mark.parametrize('typ', ['datetime64[ns]', 'datetime64[ns, GMT]',
+                                     'datetime64[ns, US/Eastern]'])
+    @pytest.mark.parametrize('op', [operator.neg,
+                                    operator.inv, operator.abs])
+    def test_datetime_raises(self, typ, op):
+        # GH#23087
+        ser = Series([pd.Timestamp('2011-01-01'), pd.Timestamp('2011-01-02'),
+                      pd.Timestamp('2011-01-03')], index=range(3), dtype=typ)
+        # pandas all raises
+        pytest.raises(TypeError, op, ser)
```
Comment (jreback) [inline pandas/tests/series/test_arithmetic.py]: can you use the other form
```
with pytest.raises(TypeError):
    op(ser)
```

**Review by jreback — CHANGES_REQUESTED** [review]
looks really good @sinhrks 

@TomAugspurger @jbrockmendel if you'd have a look. I am inclined for 0.24.0.

File: pandas/core/ops.py [hunk 14]
```diff
@@ -1435,6 +1436,82 @@ def add_flex_arithmetic_methods(cls):
...
+            return self._constructor(new_data, name=self.name,
+                                     index=self.index).__finalize__(self)
+
+    f.__name__ = op_name
+
+    return f
+
+
+def pos(values):
+    if is_datetime64_any_dtype(values):
```
Comment (TomAugspurger) [inline pandas/core/ops.py]: Why is this done here, instead of on `DatetimeArray`?

File: pandas/core/ops.py [hunk 15]
```diff
@@ -1435,6 +1436,82 @@ def add_flex_arithmetic_methods(cls):
...
+                new_data = op(self._values)
+            return self._constructor(new_data, name=self.name,
+                                     index=self.index).__finalize__(self)
+
+    f.__name__ = op_name
+
+    return f
+
+
+def pos(values):
```
Comment (TomAugspurger) [inline pandas/core/ops.py]: Should these be added to `ops._get_opstr` and  `ops._op_description`?

**Review by TomAugspurger — COMMENTED** [review]
Looks quite good. I'm also OK with 0.24.0

File: pandas/core/arrays/integer.py [hunk 16]
```diff
@@ -644,9 +644,20 @@ def integer_arithmetic_method(self, other):
         name = '__{name}__'.format(name=op.__name__)
         return set_function_name(integer_arithmetic_method, name, cls)
 
+    @classmethod
+    def _create_unary_method(cls, op):
+        def integer_unary_method(self):
+            with np.errstate(all='ignore'):
+                result = op(self._data)
+            return type(self)(result, self._mask)
```
Comment (jbrockmendel) [inline pandas/core/arrays/integer.py]: should this use _from_sequence?

File: pandas/core/ops.py [hunk 17]
```diff
@@ -1435,6 +1436,82 @@ def add_flex_arithmetic_methods(cls):
...
+# Unary
+
+def _unary_method(cls, op, special=True):
+    op_name = _get_op_name(op, special)
+
+    def f(self):
+        if self.ndim > 2:
+            with np.errstate(all='ignore'):
+                # apply op at once based on current impl
+                new_data = op(self._values)
```
Comment (jbrockmendel) [inline pandas/core/ops.py]: should this be done block/column-wise?  in particular casting to object often messes up on datetime64 columns

File: pandas/core/arrays/integer.py (hunk 16 above)
Comment (sinhrks) [reply pandas/core/arrays/integer.py]: `from_sequence` can't accept mask. Other methods (e.g. `__getitem__`) follow the same manner to make new array using the current mask.

File: pandas/core/ops.py (hunk 15 above)
Comment (sinhrks) [reply pandas/core/ops.py]: I understand `_get_opstr` is required to pass expression to `NumExpr`, and `_op_description` is to generate flex methods docstring. Is unary methods needs them?

File: pandas/tests/frame/test_arithmetic.py (hunk 11 above)
Comment (sinhrks) [reply pandas/tests/frame/test_arithmetic.py]: done. Please suggest if there is better arg name.

File: pandas/conftest.py [hunk 18]
```diff
@@ -395,6 +409,7 @@ def tz_aware_fixture(request):
 SIGNED_EA_INT_DTYPES = ["Int8", "Int16", "Int32", "Int64"]
 ALL_INT_DTYPES = UNSIGNED_INT_DTYPES + SIGNED_INT_DTYPES
 ALL_EA_INT_DTYPES = UNSIGNED_EA_INT_DTYPES + SIGNED_EA_INT_DTYPES
+ALL_NUMPY_EA_INT_DTYPES = ALL_INT_DTYPES + ALL_EA_INT_DTYPES
```
Comment (jreback) [inline pandas/conftest.py]: this is fine

File: pandas/tests/frame/test_arithmetic.py [hunk 19]
```diff
@@ -634,3 +634,171 @@ def test_arith_non_pandas_object(self):
...
+        # GH#23087
+        typ = any_numpy_ea_int_dtype
+
+        df = pd.DataFrame({'a': [1, 2, 3],
+                           'b': [1, 2, 3],
+                           'c': [1, 2, 3],
+                           'd': [1, 2, 3]}, dtype=typ)
+        result = op(df)
+
+        if isinstance(typ, str):
```
Comment (jreback) [inline pandas/tests/frame/test_arithmetic.py]: why is this needed?

File: pandas/tests/series/test_arithmetic.py [hunk 20]
```diff
@@ -170,3 +170,158 @@ def test_ser_cmp_result_names(self, names, op):
...
             result = op(ser, cidx)
             assert result.name == names[2]
+
+# ------------------------------------------------------------------
+# Unary
+
+
+class TestSeriesUnary(object):
+
+    def test_neg(self):
```
Comment (jreback) [inline pandas/tests/series/test_arithmetic.py]: where are testing other Series types?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@sinhrks also CI is failing

**Comment by jreback** [comment]
@sinhrks can you update

File: pandas/tests/series/test_arithmetic.py (hunk 20 above)
Comment (sinhrks) [reply pandas/tests/series/test_arithmetic.py]: other types are tested below.

**Comment by TomAugspurger** [comment]
Seems to be a datetime failure: https://travis-ci.org/pandas-dev/pandas/jobs/482209044#L2539

File: pandas/core/ops.py (hunk 14 above)
Comment (sinhrks) [reply pandas/core/ops.py]: I think we need a decision about this. To move `__pos__` method to `PandasArray`,  other array dtypes also need it to keep compat with `NumPy` array.

```
>>> +np.array([np.timedelta64(1)])
# array([1], dtype=timedelta64)

>>> +np.array(['a'], dtype=object)
# array(['a'], dtype=object)
```

I think prior choice is simpler:
- Make global `pos` function for all dtypes, and raise if input is `datetime64` (current impl).
- Add `__pos__` method to all `PandasArray`.

Comment (TomAugspurger) [reply pandas/core/ops.py]: By `PandasArray`, do you mean ExtensionArray?

On master, `+PandasArray` seems to work

```python
In [21]: +pd.Series([1, 2]).array
Out[21]:
<PandasArray>
[1, 2]
Length: 2, dtype: int64
```

I assume this is via `__array_ufunc__`.

For now, I would prefer that `DatetimeArray` defines a `__pos__` that raises a ValueError or TypeError.

**Comment by jreback** [comment]
This is spitting out, but *only* on the comparison on the expected result. I think these tests prob need if its a datetimelike type, coerce to i8, do the op, coerce back.

```
___ TestSeriesUnary.test_datetime_pos_raises[op0-datetime64[ns, US/Eastern]] ___
2019-01-22T03:04:55.6144005Z [gw0] linux -- Python 3.7.2 /home/vsts/miniconda3/envs/pandas-dev/bin/python
2019-01-22T03:04:55.6144263Z 
2019-01-22T03:04:55.6144496Z self = <pandas.tests.series.test_arithmetic.TestSeriesUnary object at 0x7f7c2482af60>
2019-01-22T03:04:55.6145081Z typ = 'datetime64[ns, US/Eastern]', op = <built-in function pos>
2019-01-22T03:04:55.6145309Z 
2019-01-22T03:04:55.6145805Z     @pytest.mark.parametrize('typ', ['datetime64[ns]', 'datetime64[ns, GMT]',
2019-01-22T03:04:55.6146313Z                                      'datetime64[ns, US/Eastern]'])
2019-01-22T03:04:55.6154775Z     @pytest.mark.parametrize('op', [operator.pos])
2019-01-22T03:04:55.6155197Z     def test_datetime_pos_raises(self, typ, op):
2019-01-22T03:04:55.6155359Z         # GH#23087
2019-01-22T03:04:55.6155887Z         ser = Series([pd.Timestamp('2011-01-01'), pd.Timestamp('2011-01-02'),
2019-01-22T03:04:55.6156481Z                       pd.Timestamp('2011-01-03')], index=range(3), dtype=typ)
2019-01-22T03:04:55.6156689Z         # pandas all raises
2019-01-22T03:04:55.6156870Z         with pytest.raises(TypeError):
2019-01-22T03:04:55.6157018Z             op(ser)
2019-01-22T03:04:55.6157179Z     
2019-01-22T03:04:55.6157330Z         # inconsistent with NumPy
2019-01-22T03:04:55.6157488Z >       tm.assert_numpy_array_equal(ser.values, op(ser.values))
2019-01-22T03:04:55.6158181Z E       DeprecationWarning: Applying '+' to a non-numerical array is ill-defined. Returning a copy, but in the future this will error.
2019-01-22T03:04:55.6158377Z 
```

**Comment by jreback** [comment]
@TomAugspurger if this is blocking can defer as well.

**Comment by jreback** [comment]
@sinhrks if you can move the whatsnew to 0.25.0

**Comment by jreback** [comment]
can you merge master and update

**Comment by jreback** [comment]
@pandas-dev/pandas-core if anyone would like to rebase this, very close to being ready

File: pandas/tests/series/test_arithmetic.py [hunk 21]
```diff
@@ -169,3 +169,161 @@ def test_ser_cmp_result_names(self, names, op):
...
+            result = op(ser)
+            exp = Series(operator.inv(ser.values), index=ser.index)
+            tm.assert_series_equal(result, exp)
+
+            # inconsistent with NumPy
+            with pytest.raises(TypeError):
+                op(ser.values)
+        else:
+            result = op(ser)
+            if op is operator.pos and pd.compat.numpy._is_numpy_dev:
```
Comment (WillAyd) [inline pandas/tests/series/test_arithmetic.py]: Not sure how this is going to hold up. w.r.t the linked issue:

https://github.com/numpy/numpy/pull/11450

That was milestoned for numpy 1.16 but interestingly only appears in dev with this change.

**Review by WillAyd — COMMENTED** [review]
OK @jreback got this to green. Definitely a few rough edges but lmk what you think needs to happen to get this merged

File: pandas/conftest.py [hunk 22]
```diff
@@ -223,6 +223,20 @@ def all_compare_operators(request):
     return request.param
 
 
+@pytest.fixture(params=['__pos__', '__neg__', '__inv__', '__invert__',
+                        '__abs__'])
+def all_unary_operators(request):
```
Comment (jbrockmendel) [inline pandas/conftest.py]: can this have a name reflecting the fact that these are the strings and not operator.pos, operator.neg, etc?

File: pandas/core/arrays/timedeltas.py [hunk 23]
```diff
@@ -733,6 +733,9 @@ def __rdivmod__(self, other):
         return res1, res2
 
     # Note: TimedeltaIndex overrides this in call to cls._add_numeric_methods
+    def __pos__(self):
+        return self.copy()
```
Comment (jbrockmendel) [inline pandas/core/arrays/timedeltas.py]: should this be a copy or view?

File: pandas/core/arrays/datetimes.py [hunk 24]
```diff
@@ -785,6 +785,10 @@ def _add_delta(self, delta):
         new_values = super()._add_delta(delta)
         return type(self)._from_sequence(new_values, tz=self.tz, freq='infer')
 
+    def __pos__(self):
+        raise TypeError("Unary plus expects numeric dtype, not {}"
+                        .format(self.dtype))
+
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py]: same for PeriodArray?

File: pandas/core/ops.py [hunk 25]
```diff
@@ -1571,6 +1572,76 @@ def add_flex_arithmetic_methods(cls):
     add_methods(cls, new_methods=new_methods)
 
 
+# ----------------------------------------------------------------------
+# Unary
```
Comment (jbrockmendel) [inline pandas/core/ops.py]: id be on board with an ops.unary file

**Comment by jbrockmendel** [comment]
@sinhrks can you rebase
